plugins/BEdita/Core/src/Model/Entity/ObjectEntity.php

```diff
@@ -351,4 +361,48 @@ public function isFieldTranslatable(string $name): bool
     {
         return !in_array($name, $this->notTranslatable);
     }
+
+    /**
+     * Getter for perms virtual prop.
+     * Return `null` if `Permissions` isn't in object type associations
+     * else return an array as
+     *
+     * ```
+     * [
+     *     'roles' => ['rolename_one', 'rolename_two'],
+     *     'inherited' => false,
+     * ]
+     * ```
+     *
+     * @return array|null
+     */
+    protected function _getPerms(): ?array
+    {
+        $this->loadObjectType();
+        if (!$this->object_type || !in_array('Permissions', (array)$this->object_type->associations)) {
+            return null;
+        }
+
+        $roleIds = Hash::extract((array)$this->permissions, '{n}.role_id');
+
+        if (!$this->hasProperty('permissions')) {
+            $this->getTable()->loadInto($this, ['Permissions']);
+        }
+
+        if (empty($roleIds)) {
+            return [];
+        }
+
+        $roles = TableRegistry::getTableLocator()->get('Roles')
```

Comment:
We should probably introduce some sort of cache for `Roles` - resources rarely updated and easy to invalidate cache. 
See `ObjectTypesTable` for instance, `get`, `afterSave` and `afterDelete` methods

Reply:
Maybe I would introduce cache in another specific PR

Reply:
I had though about it but if `Permissions` was already loaded we are not sure that `Roles` are loaded too, so for every `Permission` entity we have to check that roles are loaded.
We could get with something similar to:
```php
$roles = array_unique(Hash::extract((array)$this->permissions, '{n}.role.name'));

// ensure permission and roles are loaded
if (empty($roles)) {
    $this->getTable()->loadInto($this, ['Permissions.Roles']);
    $roles = array_unique(Hash::extract((array)$this->permissions, '{n}.role.name'));
}
```
that reload the permissions entities if `roles` were missing from original permissions, but maybe we could consider it as an extreme case

Reply:
I agree that we have no guarantee about Roles being loaded, but I'd treat it as a gentlemen's agreement and assume that if Permissions are loaded, their Roles are present too… at least as a first implementation, surely the snippet you wrote would be the safest way to go.

Reply:
Ok changed.... even if we are gentlemen it is better to be safe :grin: